Delegations frontoffice prototype

sauloperez opened the pull request:
#### :tophat: What? Why?

This is a very early WIP of the proposal we discussed with CoopCat last week. **Disclaimer: it will change**. At this point, we want to know which UX provides the best balance between usability for users and maintainability cost.

This relies on https://github.com/coopdevs/decidim-module-action_delegator/pull/15.

### :camera: Screenshots (optional)
So far this looks like:

![Peek 2020-09-10 11-37](https://user-images.githubusercontent.com/762088/92712488-0b468a80-f35a-11ea-8fe1-90779f451c3a.gif)

sauloperez committed:
First attempt at delegation UI

sauloperez commented on decidim-consultations/app/assets/javascripts/decidim/consultations/vote_dialog.js:
```diff
@@ -46,4 +46,18 @@ $(document).ready(function () {
+  delegationVoteButtons.click(function () {
+    voteDialog.foundation("open");
+    delegationField.val($(this).data("delegation-id"));
+  });
```
this needs to be extracted into a JS file in https://github.com/coopdevs/decidim-module-action_delegator/.

microstudi commented:
Another possibility is to add a step after changing to a delegated user.
Instead of showing the vote popup directly, reload the votation page and show in there clearly that you are going to vote for the delegated user.

sauloperez commented:
I was thinking of doing that within the vote modal itself. I'll try to do that next.

sauloperez committed:
Add warning callout in modals

The user needs to know at all times they are using a delegation.

sauloperez commented:
![Peek 2020-09-10 19-49](https://user-images.githubusercontent.com/762088/92775803-f93b0b00-f39e-11ea-9701-04365a45962a.gif)

What do you think @microstudi ? I think it's pretty clear this way. I'm just worried about yet another "extension point" we need from Decidim to render our partial or copy the views and maintain them in our module (I don't know what's worse) :see_no_evil:

sauloperez committed:
Don't show callout when no delegation is selected

microstudi commented:
Not bad @sauloperez I think we can go with that. We will see if after the tests this needs to be changed in any case!

sauloperez committed:
Move delegation's JS into module

sauloperez commented:
I'm quite happy with the solution we have found as I'm not encountering edge cases and delegations seem to be well isolated from the rest of the UI. However, I'll ask you to discuss together what the next steps should be.

sauloperez committed:
Extract delegation links as partials

sauloperez committed:
Pass consultation's parent question to partial

So we can render all possible vote button states in the delegations modal.

sauloperez commented:
A little update on how it's looking so far 
![Peek 2020-09-15 11-01](https://user-images.githubusercontent.com/762088/93189915-07907a80-f743-11ea-8e97-a39dcf2d733f.gif)

sauloperez committed:
Show delegations link also when already voted

We weren't showing the delegations link when the logged user already
voted.

sauloperez committed:
Reset the delegations modal

Because the XHR request of the vote button renders
`update_vote_button.js.erb` the vote_button partial was re-rendered and
so the delegations modal but action_delegator.js wasn't re-evaluated. As
a result, the click event listener was gone from the delegation link
making it impossible to open the delegations modal.

We followed the same approach of `update_vote_button.js.erb` and reset
the modal and the button. The Decidim's regular vote button doesn't
re-attach the click handler because the revoke button is a <button> tag
that submits a form through XHR so it doesn't need any event listener.

sauloperez committed:
Destroy a delegated vote aka. Revoke it

To do this we need to pass the granter as the author of the vote and for
that we need to find the delegation.

sauloperez commented on decidim-consultations/app/controllers/decidim/consultations/question_votes_controller.rb:
```diff
@@ -29,14 +29,40 @@ def create
+      def enforce_permission_to_unvote(subject, extra_context = {})
+        if delegation.blank?
+          enforce_permission_to :unvote, :question, question: current_question
+        else
```
this else branch is the only bit that we needed to add.

sauloperez commented:
We want with a different approach in https://github.com/coopdevs/decidim-module-action_delegator/pull/18 that makes it unnecessary to maintain a fork (thank God :joy: ). We'll keep developing these features in https://github.com/coopdevs/decidim-module-action_delegator/pull/15.
